commands_buy.go

```diff
@@ -62,5 +65,78 @@ func (b buyCmd) ToAuditEvent() types.AuditEvent {
 }
 
 func (b buyCmd) Execute() {
-	consoleLog.Warning("Not implemented: BUY")
+	abortTxIfNoAccount(b.userID)
+
+	// We want to check the most likely fail condition first. This is the case
+	// that a stock is too expensive for the buy amount. This also minimizes
+	// the time that an accout is locked.
+
+	// Get a quote for the stock
+	qr := types.QuoteRequest{
+		Stock:      b.stock,
+		UserID:     b.userID,
+		AllowCache: true,
+		ID:         b.id,
+	}
+
+	q := getQuote(qr)
+
+	// Get a fresh quote if quote is about to expire
+	quoteTTL := q.Timestamp.Add(time.Second*60).Unix() - time.Now().Unix()
+	if quoteTTL < config.QuotePolicy.UseInBuySell {
+		consoleLog.Info(" [!] Getting a fresh quote for", b.Name())
+		qr.AllowCache = false
+		q = getQuote(qr)
+	}
+
+	// Check if user can buy any stock at quote price
+	quantityToBuy, purchaseAmount := q.Price.FitsInto(b.amount)
+	consoleLog.Debugf("Want to buy %d stock", quantityToBuy)
+	if quantityToBuy < 1 {
+		abortTx(b.Name() + " Cannot buy less than one stock")
+	}
+
+	acct := accountStore[b.userID]
+	acct.Lock()
+	defer acct.Unlock()
+
+	// Check to make sure use has enough funds for buy
+	if acct.balance.ToFloat() < b.amount.ToFloat() {
```

Comment:
I agree that checking the balance covers the broadest set of cases but I don't think it's the most likely case to fail. The transaction files try to give each user boatloads of money so the accounts don't go bankrupt over the course of the run.

Doing the stock price check first allows me to push back the lock on the user account.

Does this sound reasonable @JordanVlieg ?